Implement direct mode for msvc; other improvements

shuffle2 opened the pull request:
The commit history is a bit of a mess, but I'm making a PR in hopes of getting comments on the current state.
(Ignoring the commits from my other active PRs) This PR:
* changes msvc to use something like ccache's "direct mode". This uses the input source file verbatim as input to lookup a cache entry. The cache entry has been extended to include digests of include files which were hashed when the source file was originally compiled.
* adds command-line parser for msvc, which enables handling multiple input files per buildcache invocation.
* adds support for PCH
* integrates cleanly with MSBuild's FileTracker

Currently I have completely removed the "preprocessor-style" mode for msvc, as it barely shows any speedup for me at all. Unfortunately, the `/sourceDependencies` `cl.exe` option, which my changes currently rely on, is only available on semi-recent versions of `cl.exe`. So it may be worthwhile to keep the preprocessor as a fallback path.

In any case, I welcome comments about changes to make to get this merged. It works quite well for me on my test projects, but I wouldn't be surprised if there are some tweaks needed for more general applicability.

mbitsnbites commented:
I'm sorry but this PR is just too much for me to review in a reasonable manner. It really needs to be split up into one PR per feature (a handful?). If the history was perfectly clean with independent commits, I could look at the individual commits, but that isn't the case here.

I'm particularly interested in two things (and would like to look at them independently):
* How is the multi-file (`cl.exe ... a.cpp b.cpp c.cpp ...`) support implemented? (e.g. with regards to partial hits/misses, stdout/stderr reproduction, parallelism, etc).
* How is the direct mode implemented? I'm pretty sure that there are compiler independent parts to it (that may need to be considered). There are also a number of pitfalls that need to be dealt with, such as interoperability between different nodes in a network where the install & build paths may differ, and of course pre-processor macros such as `__TIME__`. Initially, direct mode should definitely be an opt-in option, and going forward it must be possible to disable it (and fall back to pre-processor mode).

mbitsnbites commented:
I just merged a couple of your branches, so you can rebase this branch on top of the new master and get rid of a few diffs...

shuffle2 commented:
> I'm sorry but this PR is just too much for me to review in a reasonable manner. It really needs to be split up into one PR per feature (a handful?). If the history was perfectly clean with independent commits, I could look at the individual commits, but that isn't the case here.

No problem, I will organize the commits soonish. I just added FileTracker support, but there is still something slightly wrong with it I think (...i.e., still working on the PR(s) :))

> * How is the multi-file (`cl.exe ... a.cpp b.cpp c.cpp ...`) support implemented? (e.g. with regards to partial hits/misses, stdout/stderr reproduction, parallelism, etc).

The basic idea is that each file is indexed as if it had been given to buildcache individually. So the hasher state after relevant_env_var/prog_id/etc is repeatedly used as a prefix to create a final per-input-file digest, which is then used to do the cache lookup. So, where `preprocess_source` previously just returned a lump of data for input into the hasher, it now returns a lump of data per input file. Then, we attempt to lookup a cache entry for each individual file, keeping track of which ones missed the cache. If there were any misses, the list of missed files is handed back to the compiler wrapper, and it executes the compiler with only those input files. Finally, cache entries for each individual file are created.

Currently, I've cheated a bit and just removed the stdout/stderr which msvc produces in `run_for_miss`, as things work fine without it 🤷‍♂️ (to be clear, caching stdout/stderr works fine as well, it only appears confusing to a human reading the log). I guess this could be handled if needed (at least for msvc, pretty much all it does is print the input filename under normal circumstances).

> * How is the direct mode implemented? I'm pretty sure that there are compiler independent parts to it (that may need to be considered). There are also a number of pitfalls that need to be dealt with, such as interoperability between different nodes in a network where the install & build paths may differ, and of course pre-processor macros such as `__TIME__`. Initially, direct mode should definitely be an opt-in option, and going forward it must be possible to disable it (and fall back to pre-processor mode).
> 

The direct mode uses (for msvc) `hash(env("INCLUDE") || read(source-file))` as the key to lookup into the cache. The cache entry stores a `map<string, hash_t>`, which records the dependencies of that source-file and their digests, when the object was stored in the cache. This list of dependencies is produced by `cl.exe` by appending `/sourceDependencies` option to the command line which is executed in `run_for_miss`. Effectively the dependency list is just all the includes referenced directly or indirectly by the translation unit when it was compiled. All include files that are located in directories listed in the `INCLUDE` environment variable (these are all the crt, winsdk, etc. system headers) are ignored.

There is one trick to note: when the initial lookup succeeds (`INCLUDE` and the source file contents haven't changed), we still need to be able to "veto" the cache hit, depending if any of the dependencies changed. This is implemented by `filter_cache_hit`, another virtual function a wrapper may implement, which allows the wrapper to force a successfully-found cache entry to be ignored (treated as a miss).

For msvc, I've been able to make very deterministic builds (see [here](https://github.com/bazelbuild/bazel/issues/9466#issuecomment-682394297)). We could force-append some of those commandline options (or just document), but in any case, the direct mode will seemingly always have limitations similar to ccache's implementation.

For both of the above changes (multi-file and direct-mode), the behavior of the existing non-msvc wrappers should still work as before. The slight exception is that the order of inputs to the hash function has changed, and the hash function itself has been swapped for a different algorithm. I expect that existing caches would just garbage collect all the older entries over time.

shuffle2 commented:
This PR now supports msvc PCH as well. So I think I'm finally done in terms of features.

shuffle2 commented:
I don't expect to be able to merge this as-is.
